pkg/controller/operators/cleanupcsv_controller.go

```diff
@@ -0,0 +1,151 @@
+func (r *CleanupCsvReconciler) ReconcileClusterServiceVersion(ctx context.Context, req ctrl.Request) (reconcile.Result, error) {
+	// Set up a convenient log object so we don't have to type request over and over again
+	log := r.log.WithValues("request", req)
+	log.V(1).Info("reconciling csv")
+	metrics.EmitCleanupCSVReconcile(req.Namespace, req.Name)
+
+	// Fetch the CSV from the cache
+	in := &operatorsv1alpha1.ClusterServiceVersion{}
+	if err := r.Get(ctx, req.NamespacedName, in); err != nil {
+		if apierrors.IsNotFound(err) {
+			err = nil
+		} else {
+			log.Error(err, "Error finding ClusterServiceVersion")
+		}
+
+		return reconcile.Result{}, err
+	}
+
+	// Ignore copied csvs
+	if in.IsCopied() {
+		return reconcile.Result{}, nil
+	}
+
+	myFinalizerName := "olm.csv.cleanup.controller/finalizer"
+	if in.ObjectMeta.DeletionTimestamp.IsZero() {
+		// not being deleted, add finalizer if not present
+		if !controllerutil.ContainsFinalizer(in, myFinalizerName) {
+			controllerutil.AddFinalizer(in, myFinalizerName)
+			if err := r.Update(ctx, in); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+	} else {
+		// csv is being deleted
+		if controllerutil.ContainsFinalizer(in, myFinalizerName) {
+			// finalizer still present, do any clean up
+			ownerSelector := ownerutil.CSVOwnerSelector(in)
+			deleteOptions := client.DeleteAllOfOptions{
+				ListOptions: client.ListOptions{
+					LabelSelector: ownerSelector,
+				},
+			}
+			// Look for resources owned by this CSV, and delete them.
+			log.WithValues("selector", ownerSelector).Info("Cleaning up resources after CSV deletion")
+
+			err := r.Client.DeleteAllOf(ctx, &rbacv1.ClusterRoleBinding{}, &deleteOptions)
+			if client.IgnoreNotFound(err) != nil {
+				log.Error(err, "Deleting ClusterRoleBindings on CSV delete")
+				return reconcile.Result{}, err
+			}
+
+			err = r.Client.DeleteAllOf(ctx, &rbacv1.ClusterRole{}, &deleteOptions)
+			if client.IgnoreNotFound(err) != nil {
+				log.Error(err, "Deleting ClusterRoles on CSV delete")
+				return reconcile.Result{}, err
+			}
+
+			deleteOptions = client.DeleteAllOfOptions{
+				ListOptions: client.ListOptions{
+					LabelSelector: labels.SelectorFromSet(ownerutil.OwnerLabel(in, operatorsv1alpha1.ClusterServiceVersionKind)),
+				},
+			}
+
+			err = r.Client.DeleteAllOf(ctx, &admissionregistrationv1.MutatingWebhookConfiguration{}, &deleteOptions)
+			if client.IgnoreNotFound(err) != nil {
+				log.Error(err, "Deleting MutatingWebhookConfigurations on CSV delete")
+				return reconcile.Result{}, err
+			}
+
+			err = r.Client.DeleteAllOf(ctx, &admissionregistrationv1.ValidatingWebhookConfiguration{}, &deleteOptions)
+			if client.IgnoreNotFound(err) != nil {
+				log.Error(err, "Deleting ValidatingWebhookConfigurations on CSV delete")
+				return reconcile.Result{}, err
+			}
+
+			// remove our finalizer from the csv and update
+			controllerutil.RemoveFinalizer(in, myFinalizerName)
```

Comment:
If you don't set a foreground deletion policy, non-nil errors from `r.Client.DeleteAllOf()` don't necessarily mean that the objects are gone. Setting that, though, means that the handler in this controller will hang, potentially denying service if other resources are slow to clean up.

I believe all of the resources we're deleting are resources that we have informers on, so I'd suggest checking for dependent objects in the cache, and only removing the finalizer if none are present.